### PR TITLE
8292591: Experimentally add back barrier-less Java thread transitions

### PR DESCRIPTION
**Pull request by robehn**
Please consider, only implemented on x64/aarch64 linux/windows. (@TheRealMDoerr have now contributed PPC64)

On my box calling clock_gettime via JNI goes from 35ns to 28ns when enabled.

Passes t1-7 with option forced on, also passes t1-4 as is in this PR.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8292591](https://bugs.openjdk.org/browse/JDK-8292591): Experimentally add back barrier-less Java thread transitions

### Reviewers
 * [Martin Doerr](https://openjdk.org/census#mdoerr) (@TheRealMDoerr - **Reviewer**)
 * [Erik Österlund](https://openjdk.org/census#eosterlund) (@fisk - **Reviewer**) ⚠️ Review applies to [981a6cab](https://git.openjdk.org/jdk/pull/10123/files/981a6cab20b2b248d795506a51d0dcc36370eb85)
 * [Patricio Chilano Mateo](https://openjdk.org/census#pchilanomate) (@pchilano - **Reviewer**) ⚠️ Review applies to [981a6cab](https://git.openjdk.org/jdk/pull/10123/files/981a6cab20b2b248d795506a51d0dcc36370eb85)
 * [Daniel D. Daugherty](https://openjdk.org/census#dcubed) (@dcubed-ojdk - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk pull/10123/head:pull/10123` \
`$ git checkout pull/10123`

Update a local copy of the PR: \
`$ git checkout pull/10123` \
`$ git pull https://git.openjdk.org/jdk pull/10123/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 10123`

View PR using the GUI difftool: \
`$ git pr show -t 10123`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/10123.diff">https://git.openjdk.org/jdk/pull/10123.diff</a>

</details>

### CODE REVIEW

**Review by erikj79 — APPROVED** [review]
Build changes look good.

**Comment by dholmes-ora** [comment]
Can you explain the operation and placement of the system memory barriers please. It is not obvious how the elided membars are replaced with the new mechanism. "barrier-less" seems a bit of a misnomer as we're really replacing small barriers with a sledgehammer barrier in fewer places.

And does this really do enough to address the JNI overhead issue versus a true transition-less mechanism?

**Comment by robehn** [comment]
Yes, sorry. Since we had this code before I forgot adding an explanation.
The this gives back around 75% of the gained performance of transitions-less JNI calls.
Note that is against accidentally optimized JNI critical (removal was done in steps, before the final step it was faster than the original implementation, it was never intended to make it faster).
So it should be even closer the original pre-JDK 17 numbers.
But note that this applies to all JNI methods, not just some special ones.

For safepoints poll the Java thread do:
1: Store an unsafe thread state as indication that we are entering the VM.
2: Check if entrance into the VM can be performed safely.

VM Thread (or a handshaker) do:
1: Store polling word
2: Read the thread state

This must be executed in order where 1 happens before 2.
```
store unsafe thread state
store_load_barrier
load poll

store poll
store_load_barrier
load thread state
```

This patch moves store_load_barrier to the read of the thread state by the use of system memory barrier, which make sure we get program order: "guarantee that all its running thread siblings have passed through a state where all memory accesses to user-space addresses match program order"

```
store unsafe thread state
compiler_barrier
load poll

store poll
system_memory_barrier
load thread state
```

As you said this big hammer have downside since it always must be emitted before thread the thread state.
Such as:
* Using JFR sampler with short periods, or sampling many threads.
* Workload with many per seconds safepoints or handshake
Which means your overall performance may suffer and only a few special workloads should notice a difference at all.

I have not changed all transitions to elide the store_load, since they are not performance impacting and this PR was focused on native transitions.
If you think all transitions should honor this flag I can do a follow-up.

**Comment by robehn** [comment]
> Build changes look good.

Thank you!

**Review by TheRealMDoerr — COMMENTED** [review]
Nice to see usage of this OS feature. I like it more than the serialization page we have in 11u (with `-XX:-UseMembar`) and Critical JNI Natives (which still have some advantage). I have observed the same speedup for a trivial JNI call on my x86_64 machine.

**Comment by robehn** [comment]
Thanks for having a look Martin(@TheRealMDoerr), I'll see what I can do regarding the define!

File: src/hotspot/os/linux/systemMemoryBarrier_linux.cpp [hunk 1]
```diff
@@ -0,0 +1,94 @@
...
+#include "logging/log.hpp"
+#include "utilities/debug.hpp"
+#include "utilities/systemMemoryBarrier.hpp"
+
+#if defined(AMD64) || defined(AARCH64)
+
+#include <sys/syscall.h>
+// Syscall defined in kernel 4.3
+#if !defined(SYS_membarrier)
+#define SYS_membarrier 324
```
[the remark this answers is not in the document]
Comment (robehn) [reply src/hotspot/os/linux/systemMemoryBarrier_linux.cpp:36]: SYS_gettid just have large ifdef else and then an error.
I guess e.g. ppc always compile on platforms with it defined since it doesn't need to define it.

The solution with a per plattform file as zSyscall does seems to be the nicest way. But IMHO a bit out of scope for this PR.

**Comment by robehn** [comment]
There is problem using AC_CHECK_DECL during cross-compiles, so I had to remove that until that is fixed.
For now the enum value is hardcoded, it is the same on all platforms.

Regarding the defines, it seems like it's only us that needs it for some x64 builds.
If someone else needs it, they can just add it. Since that you not really happen, I this approach is fine? 

So now the membarrier is working on all Linux platforms. (but obviously local membarrier must be conditioned for any performance benefits)

File: src/hotspot/os/linux/systemMemoryBarrier_linux.cpp (hunk 1 above)
[the remark this answers is not in the document]
Comment (robehn) [reply src/hotspot/os/linux/systemMemoryBarrier_linux.cpp]: @TheRealMDoerr please have another look :)

**Review by TheRealMDoerr — CHANGES_REQUESTED** [review]
We need to support older kernel versions on PPC64 as well. And let's just make it usable on PPC64. Improves performance of trivial native calls by about 20%.
```diff
diff --git a/src/hotspot/cpu/ppc/sharedRuntime_ppc.cpp b/src/hotspot/cpu/ppc/sharedRuntime_ppc.cpp
index 5a55d8f4dc2..0902c0f35ed 100644
--- a/src/hotspot/cpu/ppc/sharedRuntime_ppc.cpp
+++ b/src/hotspot/cpu/ppc/sharedRuntime_ppc.cpp
@@ -2147,7 +2147,9 @@ nmethod *SharedRuntime::generate_native_wrapper(MacroAssembler *masm,
     Label no_block, sync;
 
     // Force this write out before the read below.
-    __ fence();
+    if (!UseSystemMemoryBarrier) {
+      __ fence();
+    }
 
     Register sync_state_addr = r_temp_4;
     Register sync_state      = r_temp_5;
diff --git a/src/hotspot/cpu/ppc/templateInterpreterGenerator_ppc.cpp b/src/hotspot/cpu/ppc/templateInterpreterGenerator_ppc.cpp
index bd4e1ce9932..260d99bef81 100644
--- a/src/hotspot/cpu/ppc/templateInterpreterGenerator_ppc.cpp
+++ b/src/hotspot/cpu/ppc/templateInterpreterGenerator_ppc.cpp
@@ -1436,7 +1436,9 @@ address TemplateInterpreterGenerator::generate_native_entry(bool synchronized) {
   __ li(R0/*thread_state*/, _thread_in_native_trans);
   __ release();
   __ stw(R0/*thread_state*/, thread_(thread_state));
-  __ fence();
+  if (!UseSystemMemoryBarrier) {
+    __ fence();
+  }
 
   // Now before we return to java we must look for a current safepoint
   // (a new safepoint can not start since we entered native_trans).
diff --git a/src/hotspot/os/linux/systemMemoryBarrier_linux.cpp b/src/hotspot/os/linux/systemMemoryBarrier_linux.cpp
index 796b596c675..66f77818ebc 100644
--- a/src/hotspot/os/linux/systemMemoryBarrier_linux.cpp
+++ b/src/hotspot/os/linux/systemMemoryBarrier_linux.cpp
@@ -35,6 +35,8 @@
 #ifndef SYS_membarrier
   #if defined(AMD64)
   #define SYS_membarrier 324
+  #elif defined(PPC64)
+  #define SYS_membarrier 365
   #else
   #error define SYS_membarrier for the arch
   #endif
```
Btw. are spaces in front of `#` ok? Some old preprocessors might have problems with that.

**Comment by robehn** [comment]
@TheRealMDoerr thanks, applied.
Do you want to be a reviewer or contributor?

Regarding space:
On Linux we had it for long time I think, see https://github.com/openjdk/jdk/blob/30d4145e2edd2d3e4cbaa8b01f7639ab53bc0991/src/hotspot/os/linux/os_linux.cpp#L306

**Comment by robehn** [comment]
@erikj79 I had to remove the ac changes, so there is now no builds changes.
We should probably remove you as reviewer, if you don't want to review the rest? :)

**Review by erikj79 — COMMENTED** [review]
Ok I'm retracting my review.

**Review by TheRealMDoerr — CHANGES_REQUESTED** [review]
Thanks for integrating my PPC64 implementation. I prefer serving as reviewer. My contribution was rather small. The implementation LGTM.
However, the test should pass in case of `MEMBARRIER_CMD_QUERY unsupported`.
Note: I believe the test should also work in product builds when adding `commands.add("-XX:+UnlockDiagnosticVMOptions");`, but I don't insist on that. I leave you free to decide.

**Comment by robehn** [comment]
Great, thanks!

Updated test, running a t1-2 on it.

File: src/hotspot/os/linux/systemMemoryBarrier_linux.cpp [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+    log_error(os)("MEMBARRIER PRIVATE_EXPEDITED unsupported");
+    return false;
+  }
+  ret = membarrier(MEMBARRIER_CMD_REGISTER_PRIVATE_EXPEDITED, 0, 0);
+  guarantee_with_errno(ret == 0, "MEMBARRIER_CMD_REGISTER_PRIVATE_EXPEDITED failed");
+  return true;
+}
+
+void LinuxSystemMemoryBarrier::emit() {
+  int s =  membarrier(MEMBARRIER_CMD_PRIVATE_EXPEDITED, 0, 0);
```
Comment (TheRealMDoerr) [inline src/hotspot/os/linux/systemMemoryBarrier_linux.cpp:83]: Extra whitespace. Feel free to remove it.

**Review by TheRealMDoerr — APPROVED** [review]
Thanks for the update! LGTM. I hope that no other platform with older linux kernel is still needed.

File: src/hotspot/os/linux/systemMemoryBarrier_linux.cpp (hunk 2 above)
Comment (robehn) [reply src/hotspot/os/linux/systemMemoryBarrier_linux.cpp]: Fixed thanks!

**Comment by robehn** [comment]
> Thanks for the update! LGTM. I hope that no other platform with older linux kernel is still needed.

Yes, thanks!

**Comment by fisk** [comment]
This looks good to me. I looked through the code to see if I can find any additional thread state filtering that would now require the system memory barrier to work reliably, but could not find any. So it would seem that you found them all.

**Review by fisk — APPROVED** [review]

**Comment by robehn** [comment]
@fisk, great, thank you!

**Comment by magicus** [comment]
I apologize that the build support was not in place to properly use the header file when present. It turned out that this was the first time we actually ran into a well-known (by me and Erik, at least ;-)) deficiency in the build system, where the split between the target and the build toolchain never were made fully complete.

File: src/hotspot/os/windows/systemMemoryBarrier_windows.cpp [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+ * accompanied this code).
+ *
+ * You should have received a copy of the GNU General Public License version
+ * 2 along with this work; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ *
```
Comment (dcubed-ojdk) [inline src/hotspot/os/windows/systemMemoryBarrier_windows.cpp:22]: Nit: extra "*" blank line in the header should be deleted.

**Review by pchilano — APPROVED** [review]
Looks good to me! I think we can avoid always emitting the memory barrier for the JFR sampler if we move it inside the thread_state_in_native() case and then re-read the state after the barrier.

File: src/hotspot/os/windows/systemMemoryBarrier_windows.cpp [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+ */
+
+#include "precompiled.hpp"
+#include "systemMemoryBarrier_windows.hpp"
+
+#include <windows.h>
+#include <processthreadsapi.h>
```
Comment (dcubed-ojdk) [inline src/hotspot/os/windows/systemMemoryBarrier_windows.cpp:29]: Should these be alpha sorted?

File: src/hotspot/os/windows/systemMemoryBarrier_windows.hpp [hunk 5]
```diff
@@ -0,0 +1,36 @@
...
+ * accompanied this code).
+ *
+ * You should have received a copy of the GNU General Public License version
+ * 2 along with this work; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ *
```
Comment (dcubed-ojdk) [inline src/hotspot/os/windows/systemMemoryBarrier_windows.hpp:22]: Nit: extra "*" blank line in the header should be deleted.

File: src/hotspot/os/windows/systemMemoryBarrier_windows.hpp [hunk 6]
```diff
@@ -0,0 +1,36 @@
...
+ */
+
+#ifndef SHARE_UTILITIES_SYSTEMMEMORYBARRIER_WINDOWS_HPP
+#define SHARE_UTILITIES_SYSTEMMEMORYBARRIER_WINDOWS_HPP
```
Comment (dcubed-ojdk) [inline src/hotspot/os/windows/systemMemoryBarrier_windows.hpp:26]: This file is not in "share/utilities" so this should be:
```
#ifndef OS_WINDOWS_SYSTEMMEMORYBARRIER_WINDOWS_HPP
#define OS_WINDOWS_SYSTEMMEMORYBARRIER_WINDOWS_HPP
```

File: src/hotspot/share/runtime/globals.hpp [hunk 7]
```diff
@@ -1294,6 +1294,9 @@ const int ObjectAlignmentInBytes = 8;
           "Delay in milliseconds for option SafepointTimeout")              \
           range(0, max_intx LP64_ONLY(/MICROUNITS))                         \
                                                                             \
+  product(bool, UseSystemMemoryBarrier, false, EXPERIMENTAL,                \
+          "Try enable system memory barrier")                               \
```
Comment (dcubed-ojdk) [inline src/hotspot/share/runtime/globals.hpp:1298]: typo: s/Try enable/Try to enable/

File: src/hotspot/share/runtime/handshake.cpp [hunk 8]
```diff
@@ -370,6 +375,10 @@ void Handshake::execute(HandshakeClosure* hs_cl, ThreadsListHandle* tlh, JavaThr
     return;
   }
 
+  if (UseSystemMemoryBarrier) {
+    SystemMemoryBarrier::emit();
+  }
```
Comment (dcubed-ojdk) [inline src/hotspot/share/runtime/handshake.cpp:380]: It's not clear to me why this emit() is here.

File: src/hotspot/share/runtime/threads.cpp [hunk 9]
```diff
@@ -552,6 +554,15 @@ jint Threads::create_vm(JavaVMInitArgs* args, bool* canTryAgain) {
   // crash Linux VM, see notes in os_linux.cpp.
   main_thread->stack_overflow_state()->create_stack_guard_pages();
 
+  if (UseSystemMemoryBarrier) {
+    if (!SystemMemoryBarrier::initialize()) {
+      // UseSystemMemoryBarrier = false;
```
Comment (dcubed-ojdk) [inline src/hotspot/share/runtime/threads.cpp:559]: Delete the commented out line? Or was this done for testing? Or???

File: src/hotspot/share/runtime/threads.cpp [hunk 10]
```diff
@@ -552,6 +554,15 @@ jint Threads::create_vm(JavaVMInitArgs* args, bool* canTryAgain) {
...
+      vm_shutdown_during_initialization("Failed to initialize request system memory barrier synchronization.");
```
Comment (dcubed-ojdk) [inline src/hotspot/share/runtime/threads.cpp:560]: typo: s/initialize request system/initialize the requested system/

File: src/hotspot/share/utilities/systemMemoryBarrier.hpp [hunk 11]
```diff
@@ -0,0 +1,60 @@
...
+ * accompanied this code).
+ *
+ * You should have received a copy of the GNU General Public License version
+ * 2 along with this work; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ *
```
Comment (dcubed-ojdk) [inline src/hotspot/share/utilities/systemMemoryBarrier.hpp:22]: Nit: extra "*" blank line in the header should be deleted.

File: test/hotspot/jtreg/runtime/handshake/HandshakeTransitionTest.java [hunk 12]
```diff
@@ -25,11 +25,13 @@
...
 import jdk.test.lib.process.ProcessTools;
 import jdk.test.lib.process.OutputAnalyzer;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.List;
 
 /*
  * @test HandshakeTransitionTest
  * @summary This does a sanity test of the poll in the native wrapper.
- * @requires vm.debug
```
Comment (dcubed-ojdk) [inline test/hotspot/jtreg/runtime/handshake/HandshakeTransitionTest.java:32]: Why was this deleted?
Have you tested this with release bits?

File: test/hotspot/jtreg/runtime/handshake/SystemMembarHandshakeTransitionTest.java [hunk 13]
```diff
@@ -0,0 +1,63 @@
...
+ * accompanied this code).
+ *
+ * You should have received a copy of the GNU General Public License version
+ * 2 along with this work; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ *
```
Comment (dcubed-ojdk) [inline test/hotspot/jtreg/runtime/handshake/SystemMembarHandshakeTransitionTest.java:22]: Nit: extra "*" blank line in the header should be deleted.

File: test/hotspot/jtreg/runtime/handshake/SystemMembarHandshakeTransitionTest.java [hunk 14]
```diff
@@ -0,0 +1,63 @@
...
+ */
+
+import jdk.test.lib.Utils;
+import jdk.test.lib.process.ProcessTools;
+import jdk.test.lib.process.OutputAnalyzer;
```
Comment (dcubed-ojdk) [inline test/hotspot/jtreg/runtime/handshake/SystemMembarHandshakeTransitionTest.java:27]: Should these be in alpha sort order?

File: test/hotspot/jtreg/runtime/handshake/SystemMembarHandshakeTransitionTest.java [hunk 15]
```diff
@@ -0,0 +1,63 @@
...
+        commands.add("-XX:+HandshakeALot");
+        commands.add("-XX:GuaranteedSafepointInterval=20");
+        commands.add("-XX:ParallelGCThreads=1");
+        commands.add("-XX:ConcGCThreads=1");
+        commands.add("-XX:CICompilerCount=2");
+        commands.add("-XX:+UnlockExperimentalVMOptions");
+        commands.add("-XX:+UseSystemMemoryBarrier");
+        commands.addAll(Arrays.asList(args));
+        commands.add("HandshakeTransitionTest$Test");
+        ProcessBuilder pb = ProcessTools.createTestJvm(commands);
```
Comment (dcubed-ojdk) [inline test/hotspot/jtreg/runtime/handshake/SystemMembarHandshakeTransitionTest.java:57]: Please make sure the new test is run with 'release' bits.

**Review by dcubed-ojdk — CHANGES_REQUESTED** [review]

File: src/hotspot/os/linux/systemMemoryBarrier_linux.cpp [hunk 16]
```diff
@@ -0,0 +1,86 @@
...
+  #endif
+#endif // SYS_membarrier
+
+// Expedited defined in kernel 4.14
+// Therefore we define it here instead of including linux/membarrier.h
+enum membarrier_cmd {
+  MEMBARRIER_CMD_QUERY                      = 0,
+  MEMBARRIER_CMD_PRIVATE_EXPEDITED          = (1 << 3),
+  MEMBARRIER_CMD_REGISTER_PRIVATE_EXPEDITED = (1 << 4),
+};
```
[the remark this answers is not in the document]
Comment (robehn) [reply src/hotspot/os/linux/systemMemoryBarrier_linux.cpp:51]: The plan is when the build system fully supports AC_CHECK_DECL, auto conf will check this for and set a define.

File: src/hotspot/os/windows/systemMemoryBarrier_windows.cpp (hunk 4 above)
Comment (robehn) [reply src/hotspot/os/windows/systemMemoryBarrier_windows.cpp:29]: On some windows setups processthreadsapi.h needs a define from windows.h. (otherwise it do not know what arch we are compiling on) (this is a bug in the windows headers)

This file only needs processthreadsapi.h according to the docs.

File: src/hotspot/share/runtime/handshake.cpp (hunk 8 above)
Comment (robehn) [reply src/hotspot/share/runtime/handshake.cpp:380]: add_operation() arms poll.
try_process() reads JavaThread state.

Since are allowing the CPU to do reordering of instructions, try_process() can see e.g. state in_native after arming the poll.
Normally this would be safe, but since we now allow reordering it's not.
By emitting the system membar we know that if we see in_native that thread will see the armed poll.

Long comment from me above, but yes it's not clear that add_operation() and try_process() are effected by this race when using system membar.

File: test/hotspot/jtreg/runtime/handshake/HandshakeTransitionTest.java (hunk 12 above)
Comment (robehn) [reply test/hotspot/jtreg/runtime/handshake/HandshakeTransitionTest.java:32]: Martin asked if we could add the diagnostics flag and make the test(s) runnable in release build.
(higher chance of a race happening in release)
So I did that, and tested it locally.

File: src/hotspot/cpu/x86/templateInterpreterGenerator_x86.cpp [hunk 17]
```diff
@@ -1081,11 +1082,12 @@ address TemplateInterpreterGenerator::generate_native_entry(bool synchronized) {
   __ movl(Address(thread, JavaThread::thread_state_offset()),
           _thread_in_native_trans);
 
-  // Force this write out before the read below
-  __ membar(Assembler::Membar_mask_bits(
-              Assembler::LoadLoad | Assembler::LoadStore |
-              Assembler::StoreLoad | Assembler::StoreStore));
-
+  if (!UseSystemMemoryBarrier) {
+    // Force this write out before the read below
```
[the remark this answers is not in the document]
Comment (robehn) [reply src/hotspot/cpu/x86/templateInterpreterGenerator_x86.cpp]: Fixed

File: src/hotspot/os/linux/systemMemoryBarrier_linux.cpp [hunk 18]
```diff
@@ -0,0 +1,86 @@
...
+  do {                                                                      \
+    int err = errno;                                                        \
+    check_type(cond, "%s: error='%s' (errno=%s)", msg, os::strerror(err),   \
+               os::errno_name(err));                                        \
+} while (false)
+
+#define guarantee_with_errno(cond, msg) check_with_errno(guarantee, cond, msg)
+
+static int membarrier(int cmd, unsigned int flags, int cpu_id) {
+  return syscall(SYS_membarrier, cmd, flags, cpu_id); // cpu_id only on >=5.10
```
[the remark this answers is not in the document]
Comment (robehn) [reply src/hotspot/os/linux/systemMemoryBarrier_linux.cpp]: Fixed

File: src/hotspot/os/linux/systemMemoryBarrier_linux.hpp [hunk 19]
```diff
@@ -0,0 +1,36 @@
...
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ *
+ */
+
+#ifndef SHARE_UTILITIES_SYSTEMMEMORYBARRIER_LINUX_HPP
+#define SHARE_UTILITIES_SYSTEMMEMORYBARRIER_LINUX_HPP
```
[the remark this answers is not in the document]
Comment (robehn) [reply src/hotspot/os/linux/systemMemoryBarrier_linux.hpp]: Fixed

File: src/hotspot/os/linux/systemMemoryBarrier_linux.hpp [hunk 20]
```diff
@@ -0,0 +1,36 @@
...
+
+#include "memory/allStatic.hpp"
+
+class LinuxSystemMemoryBarrier : public AllStatic {
+ public:
+  static bool initialize();
+  static void emit();
+};
+
+#endif // SHARE_UTILITIES_SYSTEMMEMORYBARRIER_LINUX_HPP
```
[the remark this answers is not in the document]
Comment (robehn) [reply src/hotspot/os/linux/systemMemoryBarrier_linux.hpp]: Fixed

File: src/hotspot/os/windows/systemMemoryBarrier_windows.hpp (hunk 6 above)
Comment (robehn) [reply src/hotspot/os/windows/systemMemoryBarrier_windows.hpp]: Fixed

File: src/hotspot/os/windows/systemMemoryBarrier_windows.hpp [hunk 21]
```diff
@@ -0,0 +1,36 @@
...
+
+#include "memory/allStatic.hpp"
+
+class WindowsSystemMemoryBarrier : public AllStatic {
+ public:
+  static bool initialize();
+  static void emit();
+};
+
+#endif // SHARE_UTILITIES_SYSTEMMEMORYBARRIER_WINDOWS_HPP
```
[the remark this answers is not in the document]
Comment (robehn) [reply src/hotspot/os/windows/systemMemoryBarrier_windows.hpp]: Fixed

File: src/hotspot/share/runtime/globals.hpp (hunk 7 above)
Comment (robehn) [reply src/hotspot/share/runtime/globals.hpp]: Fixed

File: src/hotspot/share/runtime/threads.cpp (hunk 9 above)
Comment (robehn) [reply src/hotspot/share/runtime/threads.cpp]: Yes, thanks, fixed.

File: src/hotspot/share/runtime/threads.cpp (hunk 10 above)
Comment (robehn) [reply src/hotspot/share/runtime/threads.cpp]: Fixed

File: test/hotspot/jtreg/runtime/handshake/SystemMembarHandshakeTransitionTest.java (hunk 14 above)
Comment (robehn) [reply test/hotspot/jtreg/runtime/handshake/SystemMembarHandshakeTransitionTest.java]: Fixed

File: test/hotspot/jtreg/runtime/handshake/SystemMembarHandshakeTransitionTest.java (hunk 15 above)
Comment (robehn) [reply test/hotspot/jtreg/runtime/handshake/SystemMembarHandshakeTransitionTest.java:57]: Yes it does.

**Comment by robehn** [comment]
> Looks good to me! I think we can avoid always emitting the memory barrier for the JFR sampler if we move it inside the thread_state_in_native() case and then re-read the state after the barrier.

Thanks! If we can make the sampler performance better when using sysmembar it would be good.

File: src/hotspot/os/linux/systemMemoryBarrier_linux.cpp [hunk 22]
```diff
@@ -0,0 +1,86 @@
...
+ * accompanied this code).
+ *
+ * You should have received a copy of the GNU General Public License version
+ * 2 along with this work; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ *
```
[the remark this answers is not in the document]
Comment (robehn) [reply src/hotspot/os/linux/systemMemoryBarrier_linux.cpp:22]: We have some variation, those files I look at same to have this style, so I'll leave it.

**Comment by robehn** [comment]
@dcubed-ojdk thanks for reviewing, I think I got all!

File: src/hotspot/os/linux/systemMemoryBarrier_linux.cpp (hunk 16 above)
[the remark this answers is not in the document]
Comment (dcubed-ojdk) [reply src/hotspot/os/linux/systemMemoryBarrier_linux.cpp:51]: Thanks for the explanation!

File: test/hotspot/jtreg/runtime/handshake/SystemMembarHandshakeTransitionTest.java (hunk 15 above)
Comment (dcubed-ojdk) [reply test/hotspot/jtreg/runtime/handshake/SystemMembarHandshakeTransitionTest.java:57]: Thanks!

File: src/hotspot/share/runtime/handshake.cpp (hunk 8 above)
Comment (dcubed-ojdk) [reply src/hotspot/share/runtime/handshake.cpp:380]: How about a comment above the emit():
// Separate the poll() in add_operation() above() from the read of JavaThread state
// in the try_process() call below.

**Review by dcubed-ojdk — APPROVED** [review]

File: src/hotspot/share/runtime/handshake.cpp (hunk 8 above)
Comment (robehn) [reply src/hotspot/share/runtime/handshake.cpp:382]: Fixed

**Comment by robehn** [comment]
Thanks @dcubed-ojdk !

**Review by TheRealMDoerr — APPROVED** [review]
Thanks for the cleanup and improved comments! LGTM.
I think omitting the fence in `transition_from_native` would be an interesting follow-up for the future as these transitions are not uncommon (and also used by JIT compiler threads).

**Comment by robehn** [comment]
Thanks again @TheRealMDoerr!

Yes, I agree that is good candidate!

**Review by dcubed-ojdk — APPROVED** [review]
